### ISSUE TITLE
BRDF Explorer project

### ISSUE THREAD

**Issue by manhnt9** [buildaworldnet/IrrlichtBAW#279]
This issue is meant to discuss software functions of BRDF Explorer.

Rendering algorithms will be discussed in #276.
- [x] ~Include [script file](https://github.com/mnauw/git-remote-hg) to clone my CEGUI branch as git submodule~
- [x] ~Apply @Florastamine's theme patch~
- [x] Push submodules online
- [x] Build CEGUI dependencies
- [x] Build CEGUI
- [x] Add new IrrlichtBAW example `13.BRDF_Explorer`
- [x] Write CEGUI management class in `ext namespace` to init/destroy/render GUI
- [x] Install CEGUI headers to INSTALL_PATH/include/irr/ext/GUI/CEGUI
- [x] Port GUI
- [x] Relocate submodules into 3rdparty
- [x] Compile cegui ext as static lib
- [x] Build dependencies on Linux instead of requiring user to install via apt
- [x] Refactor `Helpers` by @Florastamine
- [ ] Implement IrrBAW CEGUI texture loader in `ext namespace` 
- [ ] Patch CEGUI to upload sRGB (issue #263)
- [x] Render a simple model
- [x] Solve comment https://github.com/buildaworldnet/IrrlichtBAW/pull/291#discussion_r282012784, https://github.com/buildaworldnet/IrrlichtBAW/pull/291#discussion_r282012245

**Comment by manhnt9** [comment]
Note: For cloning CEGUI on Windows, Python folder should be added into PATH environment variable. For example it's OK if you can call `python` from CMD. Python 2 is required for this script to treat Mercurial repo with Git.

Asking approval from @devshgraphicsprogramming and @Crisspl, or else we can download a latest, compressed CEGUI branch before building from these [links](https://bitbucket.org/manhnt9/cegui/downloads/?tab=branches), but that's a little hard for updating CEGUI.

**Comment by devshgraphicsprogramming** [comment]
Python 2 (I have 3 but cmake is not finding it) seems to be necessary for some of our SPIR-V dependencies to not create disturbing errors when building with MSVC ( #269 ).

I'm all for Python 2 being a dependency, but please please please make CMake on Windows find it if its installed.

If possible I'd use and require Python 3 instead of 2 and solve #269 with Python 3, not 2.

CEGUI submodule can be a 3rdpart module in `./ext/cegui/3rdparty/` that is conditionally updated by CMake script only if the option is specified that we build the examples (there should really be extra cmake option switches for certain examples that rely on big extensions with dependencies).


P.S. Actually having python as a dependency is totally cool as we plan to do #92 ;)

**Comment by devshgraphicsprogramming** [comment]
> Write main code file to render the GUI, load a simple model

I don't know where you want to discuss the Functionality-but-not-strictly-Rendering aspects of the task, but generally I would expect this issue to cover the parts of the BRDF Explorer LaTeX specification that are functionality.

Speaking of which I would want the intial sample to handle model loading and have a "stub" graphics-pipline reconfiguration function which actually gets triggered (for cases when we change "heavy" material parameters, such as the parameter source [texture vs constant] and BRDF type ).

**Comment by devshgraphicsprogramming** [comment]
One more note, everything to do with the BRDFExplorer should be statically linked into the binary, we don't want a massive folder of dlls to be necessary.

You have my permission to make `irr::ext::cegui` as a static library with its own build target, **if necessary**.

Furthermore, if CEGUI needs some skin, font atlas, or other resource that comes from a file which is not "required for BRDFExplorer only". I think the defaults/built-ins should be "resource compiled". Here's a script that supposedly does so
https://github.com/vector-of-bool/cmrc

Again, the XML layout of the GUI **specific to BRDFExplorer** needs not be embedded in the executable, but the reusable parts I would resource compile into `irr::ext::cegui`.
https://github.com/vector-of-bool/cmrc

**Comment by manhnt9** [comment]
I will update my Windows environment (includes MSVC) and test if Python 3 can be used with the Mercurial script.

About the static linking, I've done it before in my engine so I can guarantee. But I suggest we do static linking for Windows only since it doesn't benefit Linux much.

**Comment by manhnt9** [comment]
About CEGUI resource, the reusable part will be stored on my CEGUI branch too so we don't need to duplicate those files in irrbaw.

Resource distribution can also be in a single .zip file (with/without password protection) like in @Florastamine's releases before. Do you really want to compile resources into exe?

**Comment by manhnt9** [comment]
After reviewing the CEGUI data, I see that there're data which isn't general enough to get accepted into CEGUI's original branch, so it won't stay in my branch too. I'll manage a separate Git repo of the CEGUI data instead since we'll both use them. Nice CMake code to have CEGUI+Data in a build process will be provided.

**Comment by devshgraphicsprogramming** [comment]
Stuff like the skins (reusable part) etc can go into `./media/` I'm not completely 100% resource compiled ;)

**Comment by manhnt9** [comment]
OK, will have them here.

Update about Python and Mercurial bridge stuff. I will configure on my machine to push from my local cegui to a git repo. So I can have single local and 2 different remotes (bitbucket hg and git). It makes the submodule much easier since I realized that in order to clone from Mercurial:
- Python must be installed
- A script must be included in irrbaw repo
- PATH environment variable (both Linux and Windows) must be modified for git to find that script

So instead I'll push cegui to a git repo, have done it before and it has perfect sync.

In short, these 2 submodules will be added:
- CEGUI
- CEGUI datafiles to be installed in irrbaw's `media` folder

Both git repos are maintained by us on GitLab.

**Comment by manhnt9** [comment]
Wasted half a day to try [Hg-Git](http://hg-git.github.io/). Met errors on both Windows & Linux. Tags failed = not pushed anything. Issue reported: https://bitbucket.org/durin42/hg-git/issues/276/failed-to-push-tags

Tried again with this: https://github.com/frej/fast-export. Now I have 2 separate cegui folders and have keep then in sync with single command.

Repos are online at: https://gitlab.com/InnerPieceOSS

Integrating submodules into example app.

**Comment by manhnt9** [comment]
![Capture](https://user-images.githubusercontent.com/11463928/57013136-46985b80-6c34-11e9-8681-40b6add90dcd.PNG)

**Comment by manhnt9** [comment]
CEGUI dependencies now can be downloaded, built and installed to CEGUI. Will write build script for CEGUI tomorrow.

**Comment by manhnt9** [comment]
Base project is created. I will spend the day mostly for studying GUI implementation by @Florastamine.

**Comment by manhnt9** [comment]
I ported the main window GUI, but fails to compile (https://github.com/buildaworldnet/IrrlichtBAW/issues/294) so I'm waiting for the fix.
Next is to port color picker and finalize UI then we can start rendering. Hopefully in this week.
P/S: Doing some overall maintenance and test more on both platforms actually.

**Comment by manhnt9** [comment]
I meet [this](https://stackoverflow.com/questions/3336995/git-will-not-init-sync-update-new-submodules) on Windows, git doesn't recognize new submodule.

**Comment by manhnt9** [comment]
Thought it was a folder so I ignored submodule before. Solved that issue.

**Comment by devshgraphicsprogramming** [comment]
> Install CEGUI headers to INSTALL_PATH/include/irr/ext/CEGUI

I thought you meant only the CEGUI extension's headers, i.e. the headers which are supposed to be public in `irr/ext/cegui` (required for compilation of binaries and libraries using it)

**Comment by manhnt9** [comment]
ext will only help you set up. In order to develop GUI, you'll have to use CEGUI API too.

**Comment by devshgraphicsprogramming** [comment]
ok, but then:
`INSTALL_PATH/include/irr/ext/CEGUI` is for your own headers
`INSTALL_PATH/include/irr/ext/CEGUI/some/other/path` is for cegui's headers

**Comment by manhnt9** [comment]
Some initial render, still has layout bug of my port I guess. Will have more updates soon.

![cap](https://user-images.githubusercontent.com/11463928/57592911-3dc45580-7563-11e9-916a-ba7c78afef04.png)

**Comment by manhnt9** [comment]
Ah the layout is fine, just lacking windows because those are still not ported.
Great!
Also task list is updated.

**Comment by manhnt9** [comment]
This could be a bug actually, I double-checked it in CEED.
It should be like this (screenshot from CEED). Gonna have fully working main layout first before continuing.
What should be done next is to compile the old GUI demo, disable some code and compare with irrbaw app.

![Screenshot_2019-05-13_16-56-19](https://user-images.githubusercontent.com/11463928/57613022-2061ac80-75a0-11e9-9ebe-38f8654cded2.png)

**Comment by manhnt9** [comment]
Fixed by @devshgraphicsprogramming 's suggestion on Discord. This is about OpenGL/Vulkan NDC.
Mouse input works, will test key input too.

![Screenshot_2019-05-14_22-34-55](https://user-images.githubusercontent.com/11463928/57711387-cdb7eb80-7698-11e9-87b2-a8964031af88.png)

**Comment by devshgraphicsprogramming** [comment]
@manhnt9 does this help you with your Saga3D renderer?

**Comment by manhnt9** [comment]
Oooh, absolutely. Thanks for reminding me about that. I don't remember solving any NDC issue and also my GUI is disappearing somehow. Gonna revisit it soon.

**Comment by manhnt9** [comment]
Mouse and key are working fully.

**Comment by manhnt9** [comment]
All Sliders, Click to browse events and File Dialogs are working.

![Screenshot_2019-05-15_15-45-04](https://user-images.githubusercontent.com/11463928/57761899-ad377200-7728-11e9-9916-87de5e0bd733.png)

**Comment by manhnt9** [comment]
Dropdown works.
![Screenshot_2019-05-15_16-12-45](https://user-images.githubusercontent.com/11463928/57763769-4ddb6100-772c-11e9-87c6-534c08fb0c6b.png)

**Comment by manhnt9** [comment]
GUI is 99% done. Next step is to re-test as much as possible and do the rendering.
Then connect GUI data to graphics pipeline.

**Comment by manhnt9** [comment]
Current bug:
Having a crash and this error message when clicking to Choose color for testing color picker
15/05/2019 16:35:58 (Error)    CEGUI::UnknownObjectException in function 'CEGUI::NamedElement* CEGUI::NamedElement::getChildElement(const String&) const' (/home/manh/IrrlichtBAW/3rdparty/CEGUI/cegui/cegui/src/NamedElement.cpp:150) : The Element object referenced by 'auto_oldcolourrect' is not attached to Element at 'UIRoot/__cewin_uid_2'.

**Comment by manhnt9** [comment]
All GUI controls are tested, working well. Gonna test on Windows.

**Comment by manhnt9** [comment]
I'm gonna refactor the build process for CEGUI first then solve pending comments later.

**Comment by manhnt9** [comment]
We have perfect build now, Debug/Release works, CEGUI projects are built before building IrrExtCEGUI.
Also found a bug about disappearing cursor in file dialog and fixed it. App has 2 overlapping cursors, CEGUI cursor and native cursor. Native cursor is now only visible for file dialog.

![Capture](https://user-images.githubusercontent.com/11463928/58013814-a3937d00-7b21-11e9-9aad-1decbefb8844.PNG)

**Comment by manhnt9** [comment]
Base project is considered done to me. From now we can add new code consistently.
CEGUI's dependencies will be upgraded later when the Ext is widely used in examples then I'll change it to submodule instead of download and build from .zip. Hope that by the time CEGUI is already moved to Git too.

**Comment by devshgraphicsprogramming** [comment]
so do we close this issue?

**Comment by manhnt9** [comment]
Still have sRGB tasks left.

**Comment by manhnt9** [comment]
Next week I'll merge with the loaders, write the image codec for CEGUI, make CEGUI upload textures in SRGB format.

**Comment by devshgraphicsprogramming** [comment]
Maybe we should remove the following from the checklist

> More rendering features based on #276

so at least we can close issues

**Comment by devshgraphicsprogramming** [comment]
if IrrBaW will provide texture loading abilities to CEGUI can we kick out STB, etc?

**Comment by devshgraphicsprogramming** [comment]
can you guys manage to fix the crash when the titlebar of the color picker is double-clicked?
(or at least disable any sort of reponse to a double click)

**Comment by devshgraphicsprogramming** [comment]
@Crisspl would it make sense to get around our texture flipping issues with:
- [ ] Implement IrrBAW CEGUI texture loader in ext namespace
?


Then we kick out STB and not build the CEGUI texture loading lib.


@Crisspl and @Przemog1 at your leisure:
- [ ] Patch CEGUI to upload/use sRGB textures (issue #263) right now its rendering with sRGB "reinterpret_cast'ed" as linear, outputting into an SRGB framebuffer . I don't know how much easier the above point would make that.
- [ ] Patch the CEGUI color picker to not fucking crash when double clicked

The last two things would be modifcations of the CEGUI submodules which would have to be diff'ed (with your favourite diff tool) into a .patch file that we submit to @manhnt9's branch of CEGUI (since his git repo is kind-of a mirror of a mercurial repo, yes CEGUI people use Hg)

**Comment by Crisspl** [comment]
>would it make sense to get around our texture flipping issues with:
>Implement IrrBAW CEGUI texture loader in ext namespace

You'd have to elaborate. 
The only solution im almost sure will work is to separate CEGUI textures and irrbaw textures on the level of GL objects. Still i can't be 100% sure because it also became apparent that this weird behavior affects also textures that cegui doesnt even know about (but simultaneusly there are some shared between irrbaw and cegui). However it worked (textures wasnt flipped) when there was only GUI - no irrbaw textures were shared with cegui nor created at all

**Comment by devshgraphicsprogramming** [comment]
To be fair I don't even know how on earth CEGUI could flip textures in the main render window.

It does make most sense for CEGUI textures and our own textures to be separate objects.
(Unless we provided our own GUI Renderer to replace OpenGL core one)
